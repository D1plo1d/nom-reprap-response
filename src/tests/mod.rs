use insta::assert_debug_snapshot;
use std::collections::BTreeMap;
use colored::Colorize;

use eyre::{
    eyre,
    // Context as _,
};

pub use super::{
    parse_response,
    Response,
};

fn responses_for(v: &str) -> eyre::Result<Vec<Response>> {
    std::iter::repeat(true)
        .scan(v, |acc, _| {
            let mut try_parse = move || {
                if acc.len() == 0 {
                    return Ok(None)
                }

                let (
                    remaining,
                    (parsed, response)
                ) = parse_response(acc)
                    .map_err(|err| {
                        eyre!(
                            "[{}]\n Error: {:?}. Input: {:?}\n",
                            "ERROR".red(),
                            err,
                            acc,
                        )
                    })?;

                if let Response::Unknown = response {
                    Err(eyre!(
                        "[{}]\nParsed: {:?} . Input: {:?}\n",
                        "UNKNOWN RESPONSE".red(),
                        parsed,
                        acc,
                    ))?;
                };

                *acc = remaining;

                Ok(Some(response))
            };

            try_parse().transpose()
        })
        .collect::<eyre::Result<Vec<Response>>>()
}

fn snapshot_test_responses(data: &str) -> eyre::Result<()> {
    let data: BTreeMap<String, BTreeMap<String, String>> = toml::from_str(data)?;

    type SectionResponseTree = BTreeMap<String, Vec<Response>>;
    type Snapshot = BTreeMap<String, SectionResponseTree>;

    let responses: eyre::Result<Snapshot> = data
        .iter()
        .map(|(section_title, scenarios)| {
            let section_title = section_title.clone();

            let section_responses = scenarios.clone()
                .iter()
                .map(|(k, v)| {
                    let title = format!("{}::{}", &section_title, k);

                    print!("\nTesting {:?} ", title);
                    let res = responses_for(&v)?;

                    println!("[{}]", "DONE".green());
                    println!("Got: {:?}\n", res);

                    Ok((k.to_owned(), res))
                })
                .collect::<eyre::Result<SectionResponseTree>>()?;

            Ok((section_title, section_responses))
        })
        .collect();

    assert_debug_snapshot!(responses?);

    Ok(())
}

#[test]
fn ender3_marlin_2019_firmware() -> eyre::Result<()> {
    let data = include_str!("data/ender3_marlin_2019_firmware.toml");
    snapshot_test_responses(data)
}

#[test]
fn ultimaker2_marlin_dbg_2019_firmware() -> eyre::Result<()> {
    let data = include_str!("data/ultimaker2_marlin_dbg_2019_firmware.toml");
    snapshot_test_responses(data)
}
