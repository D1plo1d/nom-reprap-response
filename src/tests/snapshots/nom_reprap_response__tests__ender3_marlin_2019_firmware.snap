---
source: src/tests/mod.rs
expression: responses?

---
{
    "errors": {
        "min_temp": [
            Feedback(
                ActualTemperatures(
                    [
                        (
                            "e0",
                            21.95,
                        ),
                        (
                            "b",
                            22.0,
                        ),
                    ],
                ),
            ),
            Error(
                "0\n: Extruder switched off. MINTEMP triggered !",
            ),
            Error(
                "Printer stopped due to errors. Fix the error and use M999 to restart. (Temperature is reset. Set it after restarting)",
            ),
            Ok(
                None,
            ),
        ],
        "unknown_command": [
            Echo(
                "Unknown command: \"N27 m23 file.txt*119\"",
            ),
            Ok(
                None,
            ),
        ],
    },
    "firmware": {
        "m115_firmware_info": [
            Echo(
                "FIRMWARE_NAME:Marlin V1; Sprinter/grbl mashup for gen6 FIRMWARE_URL:http://www.mendel-parts.com PROTOCOL_VERSION:1.0 MACHINE_TYPE:www.creality3d.cn EXTRUDER_COUNT:1 UUID:00000000-0000-0000-0000-000000000000",
            ),
            Ok(
                None,
            ),
        ],
    },
    "greeting": {
        "greeting_with_sd_card": [
            Greeting,
            Echo(
                "External Reset",
            ),
            Greeting,
            Echo(
                "Last Updated: Sep  3 2018 15:45:59 | Author: (Ender3)",
            ),
            Debug(
                "Compiled: Sep  3 2018",
            ),
            Echo(
                "Free Memory: 9679  PlannerBufferBytes: 1232",
            ),
            Echo(
                "Hardcoded Default Settings Loaded",
            ),
            Echo(
                "SD card ok",
            ),
            Debug(
                "Init power off infomation.",
            ),
            Debug(
                "size: \n591",
            ),
        ],
        "greeting_without_sd_card": [
            Greeting,
            Echo(
                "External Reset",
            ),
            Greeting,
            Echo(
                "Last Updated: Sep  3 2018 15:45:59 | Author: (Ender3)",
            ),
            Debug(
                "Compiled: Sep  3 2018",
            ),
            Echo(
                "Free Memory: 9679  PlannerBufferBytes: 1232",
            ),
            Echo(
                "Hardcoded Default Settings Loaded",
            ),
            Echo(
                "SD init fail",
            ),
            Echo(
                "SD init fail",
            ),
        ],
    },
    "m20_list_sd_card": {
        "m20_sucess": [
            Ok(
                None,
            ),
        ],
    },
    "m21_init_sd_card": {
        "m21_no_sd_card": [
            Echo(
                "SD init fail",
            ),
            Echo(
                "SD init fail",
            ),
            Ok(
                None,
            ),
        ],
        "m21_success": [
            Echo(
                "SD card ok",
            ),
            Debug(
                "Init power off infomation.",
            ),
            Debug(
                "size: \n591",
            ),
        ],
    },
    "m23_select_sd_file": {
        "m23_no_sd_card": [
            Error(
                "Failed to open sd card file. Check that SD card is inserted and enabled in firmware.",
            ),
        ],
        "m23_success": [
            Debug(
                "File opened: file.tx Size: 17\nFile selected",
            ),
            Ok(
                None,
            ),
        ],
    },
    "m24_start_or_resume_sd_print": {
        "m24_no_file_selected": [
            Debug(
                "File(bin) deleted.",
            ),
            Ok(
                None,
            ),
        ],
        "m24_success": [
            Debug(
                "Deletion(bin) failed.",
            ),
            Ok(
                None,
            ),
            Feedback(
                Positions(
                    Positions {
                        target_positions: Some(
                            [
                                (
                                    "x",
                                    0.0,
                                ),
                                (
                                    "y",
                                    0.0,
                                ),
                                (
                                    "z",
                                    0.0,
                                ),
                                (
                                    "e0",
                                    0.0,
                                ),
                            ],
                        ),
                        actual_positions: [
                            (
                                "x",
                                0.0,
                            ),
                            (
                                "y",
                                0.0,
                            ),
                            (
                                "z",
                                0.0,
                            ),
                        ],
                    },
                ),
            ),
            Ok(
                Some(
                    ActualTemperatures(
                        [
                            (
                                "e0",
                                26.3,
                            ),
                            (
                                "b",
                                25.9,
                            ),
                            (
                                "e0",
                                26.3,
                            ),
                        ],
                    ),
                ),
            ),
            Ok(
                Some(
                    SDPrintComplete,
                ),
            ),
        ],
    },
    "m28_start_sd_write": {
        "m28_no_sd_card": [
            Error(
                "Failed to open sd card file. Check that SD card is inserted and enabled in firmware.",
            ),
        ],
        "m28_success": [
            Ok(
                Some(
                    StartSDWrite(
                        StartSDWrite {
                            filename: "file.txt",
                        },
                    ),
                ),
            ),
            Ok(
                None,
            ),
        ],
    },
    "m29_stop_sd_write": {
        "m29_file_open": [
            Ok(
                Some(
                    SDWriteComplete,
                ),
            ),
        ],
        "m29_no_file_open": [
            Ok(
                None,
            ),
        ],
    },
    "m30_delete_sd_file": {
        "m30_no_file": [
            Error(
                "Deletion failed, File: teg.gcod.",
            ),
        ],
        "m30_success": [
            Debug(
                "File deleted:file.txok",
            ),
        ],
    },
    "movement_gcodes": {
        "g1": [
            Ok(
                None,
            ),
        ],
        "g28": [
            Ok(
                None,
            ),
        ],
        "g90": [
            Ok(
                None,
            ),
        ],
        "g91": [
            Ok(
                None,
            ),
        ],
    },
    "polling_mcodes": {
        "m105": [
            Ok(
                Some(
                    ActualTemperatures(
                        [
                            (
                                "e0",
                                25.9,
                            ),
                            (
                                "b",
                                25.5,
                            ),
                            (
                                "e0",
                                25.9,
                            ),
                        ],
                    ),
                ),
            ),
        ],
        "m114": [
            Feedback(
                Positions(
                    Positions {
                        target_positions: Some(
                            [
                                (
                                    "x",
                                    100.0,
                                ),
                                (
                                    "y",
                                    0.0,
                                ),
                                (
                                    "z",
                                    0.0,
                                ),
                                (
                                    "e0",
                                    0.0,
                                ),
                            ],
                        ),
                        actual_positions: [
                            (
                                "x",
                                47.6,
                            ),
                            (
                                "y",
                                0.0,
                            ),
                            (
                                "z",
                                0.0,
                            ),
                        ],
                    },
                ),
            ),
            Ok(
                None,
            ),
        ],
        "m155_temperature_auto_report": [
            Ok(
                None,
            ),
        ],
    },
    "temperature_control": {
        "m109": [
            Feedback(
                ActualTemperatures(
                    [
                        (
                            "e0",
                            24.8,
                        ),
                    ],
                ),
            ),
            Feedback(
                ActualTemperatures(
                    [
                        (
                            "e0",
                            25.0,
                        ),
                    ],
                ),
            ),
        ],
    },
}
